crates/sui-framework/packages/move-stdlib/sources/vector.move

```diff
@@ -158,4 +158,110 @@ module std::vector {
+        v.destroy_empty();
+    }
+
+    /// Perform an action `f` on each element of the vector `v`. The vector is not modified.
+    public macro fun do_ref<$T>($v: &vector<$T>, $f: |&$T|) {
+        let v = $v;
+        v.length().do!(|i| $f(&v[i]))
+    }
+
+    /// Perform an action `f` on each element of the vector `v`. The function `f` takes a mutable
+    /// reference.
+    public macro fun do_mut<$T>($v: &mut vector<$T>, $f: |&mut $T|) {
+        let v = $v;
+        v.length().do!(|i| $f(&mut v[i]))
+    }
+
+    /// Map the vector `v` to a new vector by applying the function `f` to each element.
+    /// Preserve the order of elements in the vector, first is called first.
+    public macro fun map<$T, $U>($v: vector<$T>, $f: |$T| -> $U): vector<$U> {
+        let v = $v;
+        let mut r = vector[];
+        v.do!(|e| r.push_back($f(e)));
+        r
+    }
+
+    /// Map the vector `v` to a new vector by applying the function `f` to each element.
+    /// Preserve the order of elements in the vector, first is called first.
+    public macro fun map_ref<$T, $U>($v: &vector<$T>, $f: |&$T| -> $U): vector<$U> {
+        let v = $v;
+        let mut r = vector[];
+        v.do_ref!(|e| r.push_back($f(e)));
+        r
+    }
+
+    /// Filter the vector `v` by applying the function `f` to each element.
+    /// Return a new vector containing only the elements for which `f` returns `true`.
+    public macro fun filter<$T: drop>($v: vector<$T>, $f: |&$T| -> bool): vector<$T> {
+        let v = $v;
+        let mut r = vector[];
+        v.do!(|e| if ($f(&e)) r.push_back(e));
+        r
+    }
+
+    /// Split the vector `v` into two vectors by applying the function `f` to each element.
+    /// Return a tuple containing two vectors: the first containing the elements for which `f` returns `true`,
+    /// and the second containing the elements for which `f` returns `false`.
+    public macro fun partition<$T>($v: vector<$T>, $f: |&$T| -> bool): (vector<$T>, vector<$T>) {
+        let v = $v;
+        let mut r1 = vector[];
+        let mut r2 = vector[];
+        v.do!(|e| if ($f(&e)) r1.push_back(e) else r2.push_back(e));
+        (r1, r2)
+    }
+
+    /// Reduce the vector `v` to a single value by applying the function `f` to each element.
+    /// Similar to `fold_left` in Rust and `reduce` in Python and JavaScript.
+    public macro fun fold<$T, $Acc>($v: vector<$T>, $init: $Acc, $f: |$Acc, $T| -> $Acc): $Acc {
+        let v = $v;
+        let mut acc = $init;
+        v.do!(|e| acc = $f(acc, e));
+        acc
+    }
+
+    /// Whether any element in the vector `v` satisfies the predicate `f`.
+    /// Does not preserve the order of elements in the vector.
+    public macro fun any<$T>($v: &vector<$T>, $f: |&$T| -> bool): bool {
+        let v = $v;
+        let mut i = 0;
+        let len = v.length();
+        while (i < len) {
+            if ($f(&v[i])) return true;
+            i = i + 1;
+        };
+        false
```

Comment:
```suggestion
        'a: {
            v.length().do!(|i| if ($f(&v[i])) return 'a true);
            false 
        }
```

Reply:
@cgswords we should probably resurrect some discussions about break/continue/return with macros
Its a good example to work with

Reply:
I'm not sure if I like this one specifically...

Reply:
But maybe I'll get used to it. The idea of wrapping the contents into a labeled block is meh.

Reply:
I realized this should really be 
```
'a: {
    v.do_ref!(|x| if ($f(x)) return 'a true);
    false 
}

I would suggest doing this for now mostly to show off the feature. 
Long term I think we would prefer something like 
```
v.do_ref!(|x| if ($f(x)) return 'any true);
false 
```

Reply:
Did it, thanks!